File: app/src/main/java/im/tox/antox/callbacks/AntoxOnFileControlCallback.java
package im.tox.antox.callbacks;

import android.content.Context;
import android.util.Log;

import java.nio.ByteBuffer;

import im.tox.antox.data.AntoxDB;
import im.tox.antox.tox.ToxSingleton;
import im.tox.antox.utils.AntoxFriend;
import im.tox.jtoxcore.ToxFileControl;
import im.tox.jtoxcore.callbacks.OnFileControlCallback;

public class AntoxOnFileControlCallback implements OnFileControlCallback<AntoxFriend> {

    private static final String TAG = "OnFileControlCallback";
    private Context ctx;
    ToxSingleton toxSingleton = ToxSingleton.getInstance();

    public AntoxOnFileControlCallback(Context ctx) { this.ctx = ctx; }

    public void execute(AntoxFriend friend, boolean sending, int fileNumber, ToxFileControl control_type, byte[] data) {
        Log.d(TAG, "execute, control type: " + control_type.name() + " sending: " + sending);
        if (control_type.equals(ToxFileControl.TOX_FILECONTROL_FINISHED) && !sending) {
            Log.d(TAG, "TOX_FILECONTROL_FINISHED");
            toxSingleton.fileFinished(friend.getId(), fileNumber, ctx);
        }
        if (control_type.equals(ToxFileControl.TOX_FILECONTROL_ACCEPT) && sending) {
            AntoxDB antoxDB = new AntoxDB(ctx);
            int id = antoxDB.getFileId(friend.getId(), fileNumber);
            if (id != -1) {
                if (!toxSingleton.fileStatusMap.containsKey(id) || toxSingleton.fileStatusMap.get(id).equals(ToxSingleton.FileStatus.REQUESTSENT)) {
                    antoxDB.fileTransferStarted(friend.getId(), fileNumber);
                }
            }
            antoxDB.close();
            if (id != -1) {
                if (!toxSingleton.fileStatusMap.containsKey(id) || toxSingleton.fileStatusMap.get(id).equals(ToxSingleton.FileStatus.REQUESTSENT)) {
                    antoxDB.fileTransferStarted(friend.getId(), fileNumber);
                    toxSingleton.updatedMessagesSubject.onNext(true);
                    toxSingleton.sendFileData(friend.getId(), fileNumber, 0, ctx);
                } else if (toxSingleton.fileStatusMap.get(id).equals(ToxSingleton.FileStatus.PAUSED)) {
                    toxSingleton.sendFileData(friend.getId(), fileNumber, toxSingleton.getProgress(id), ctx);
                }
            }
        }
        if (control_type.equals(ToxFileControl.TOX_FILECONTROL_RESUME_BROKEN) && sending) {
            try {
                toxSingleton.jTox.fileSendControl(friend.getFriendnumber(), true, fileNumber, ToxFileControl.TOX_FILECONTROL_ACCEPT.ordinal(), new byte[0]);
            } catch (Exception e) {
                e.printStackTrace();
            }
            toxSingleton.sendFileData(friend.getId(), fileNumber, (int) ByteBuffer.wrap(data).getLong(), ctx);
        }
        if (control_type.equals(ToxFileControl.TOX_FILECONTROL_PAUSE) && sending) {
            AntoxDB antoxDB = new AntoxDB(ctx);
            int id = antoxDB.getFileId(friend.getId(), fileNumber);
            antoxDB.close();
            if (id != -1) {
                toxSingleton.fileStatusMap.put(id, ToxSingleton.FileStatus.PAUSED);
            }
        }
        if (control_type.equals(ToxFileControl.TOX_FILECONTROL_KILL) && sending) {
            toxSingleton.cancelFile(friend.getId(), fileNumber, ctx);
        }
    }
}

File: app/src/main/scala/im/tox/antox/callbacks/AntoxOnFileControlCallback.scala
package im.tox.antox.callbacks

import android.content.Context
import android.util.Log
import im.tox.antox.callbacks.AntoxOnFileControlCallback._
import im.tox.antox.data.State
import im.tox.antox.tox.{Reactive, ToxSingleton}
import im.tox.antox.transfer.FileStatus
import im.tox.tox4j.core.callbacks.FileControlCallback
import im.tox.tox4j.core.enums.ToxFileControl

object AntoxOnFileControlCallback {

  private val TAG = "OnFileControlCallback"
}

class AntoxOnFileControlCallback(private var ctx: Context) extends FileControlCallback {
  
  override def fileControl(friendNumber: Int, fileNumber: Int, control: ToxFileControl): Unit = {
      Log.d(TAG, "control type: " + control.name())
      val mTransfer = State.transfers.get(ToxSingleton.getAntoxFriend(friendNumber).get.getKey, fileNumber)
      mTransfer match {
        case Some(t) =>
          (control, t.status) match {
            case (ToxFileControl.RESUME, FileStatus.REQUESTSENT) =>
              Log.d(TAG, "fileTransferStarted")
              State.transfers.fileTransferStarted(t.key, t.fileNumber, ctx)
            case (ToxFileControl.RESUME, FileStatus.PAUSED) =>
              Log.d(TAG, "fileTransferResumed")
              State.transfers.fileTransferStarted(t.key, t.fileNumber, ctx)
            case (ToxFileControl.PAUSE, _) =>
              Log.d(TAG, "pauseFile")
              State.transfers.pauseFile(t.id, ctx)
            case (ToxFileControl.CANCEL, _) =>
              Log.d(TAG, "cancelFile")
              State.transfers.cancelFile(t.key, t.fileNumber, ctx)
            case _ =>
              Log.d(TAG, "not matched: " + control + ", " + t.status)

          }
        case None => Log.d(TAG, "Transfer not found")
      }
      Reactive.updatedMessages.onNext(true)
    }
}
